### ISSUE TITLE
Incorrect Magnitude Calculation for DoRA Linear Layers (Violates DoRA Paper Methodology)

### ISSUE THREAD

**Issue by arcteryox** [huggingface/peft#2348]
### **Description**  
The current `DoraLinearLayer` incorrectly computes weight magnitude norms **per input channel** instead of **per output channel**, violating the methodology outlined in the [DoRA paper (Section 3.1)](https://arxiv.org/abs/2402.09353). This leads to degraded performance for linear layers (e.g., in LLMs).

---

#### **Affected Code**:  
`peft/tuners/lora/dora.py` → `DoraLinearLayer.get_weight_norm`  
```python
def get_weight_norm(self, weight, lora_weight, scaling):
    weight = transpose(weight, self.fan_in_fan_out)  # ❌ Transposes to [in_features, out_features]
    weight = weight + scaling * lora_weight
    weight_norm = torch.linalg.norm(weight, dim=1)   # Norm over input channels (dim=1)
    return weight_norm
```

#### **Problem**:  
- For a linear layer with weight shape `[out_features, in_features]`, transposing to `[in_features, out_features]` causes `dim=1` to represent **input channels**, not output channels.  
- This contradicts the DoRA paper’s requirement to compute magnitude **per output channel** (rows of the weight matrix).

---

### **Steps to Reproduce**  
1. Initialize a DoRA-linear layer:  
```python
base_layer = nn.Linear(10, 5)  # out_features=5, in_features=10
dora_layer = DoraLinearLayer(fan_in_fan_out=False)
```

2. Check weight norm dimensions:  
```python
weight = base_layer.weight  # Shape [5, 10]
lora_weight = torch.randn(5, 10)  # Simulate LoRA delta
norm = dora_layer.get_weight_norm(weight, lora_weight, scaling=1.0)
print(norm.shape)  # Outputs [10] (input channels) instead of [5] (output channels)
```

---

### **Expected vs Actual Behavior**  
| Expected (Per Paper) | Actual (Current Code) |  
|-----------------------|-----------------------|  
| Norms computed over **output channels** (`out_features`). | Norms computed over **input channels** (`in_features`). |  

---

### **Proposed Fix**  
Remove the transpose and compute norms over `dim=1` directly:  
```python
def get_weight_norm(self, weight, lora_weight, scaling):
    # Remove transpose - work directly with [out_features, in_features]
    weight = weight + scaling * lora_weight
    weight_norm = torch.linalg.norm(weight, dim=1)  # ✅ Norm over output channels (dim=1)
    return weight_norm
```

#### **Impact of Fix**:  
- Aligns with DoRA paper’s methodology for linear layers.  
- Convolutional layers (e.g., `DoraConv2dLayer`) are unaffected and already correct.  

---

### **Additional Context**  
1. **Paper Reference**:  
   - Section 3.1 defines magnitude as the L2 norm of **rows** (output channels) for linear layers.  
   - Example: For weight matrix `W ∈ ℝ^{d×k}`, magnitude `m_j = ||W_j||_2` (row-wise norm).  

2. **Why This Matters**:  
   - Magnitude scaling is critical for DoRA’s ability to decouple direction and magnitude updates.  
   - Incorrect scaling invalidates the method’s theoretical guarantees and reduces performance (e.g., on LLM fine-tuning tasks).

---

### **Verification**  
After applying the fix:  
```python
print(norm.shape)  # Now outputs [5] (correct for out_features=5)
``` 

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### **Steps to Reproduce**  
1. Initialize a DoRA-linear layer:  
```python
base_layer = nn.Linear(10, 5)  # out_features=5, in_features=10
dora_layer = DoraLinearLayer(fan_in_fan_out=False)
```

2. Check weight norm dimensions:  
```python
weight = base_layer.weight  # Shape [5, 10]
lora_weight = torch.randn(5, 10)  # Simulate LoRA delta
norm = dora_layer.get_weight_norm(weight, lora_weight, scaling=1.0)
print(norm.shape)  # Outputs [10] (input channels) instead of [5] (output channels)
```

### **Expected vs Actual Behavior**  
| Expected (Per Paper) | Actual (Current Code) |  
|-----------------------|-----------------------|  
| Norms computed over **output channels** (`out_features`). | Norms computed over **input channels** (`in_features`). |

**Comment by BenjaminBossan** [comment]
Thank you for the detailed write up. This part of the paper/code has caused some confusion in the past. Indeed, IIRC, I also implemented it along the other axis, as you suggest. However, the paper author nbasyl thankfully reviewed the PR (#1474) and therefore we believe it should be correct.

There was a similar discussion on the LyCORIS repo, for instance check [my comment there](https://github.com/KohakuBlueleaf/LyCORIS/issues/216#issuecomment-2374087115). At the end of the day, it appears that either axis works, I'm not sure why.

**Comment by arcteryox** [comment]
Thank you for pointing me to the related discussions.

After closely reviewing the [original DoRA implementation](https://github.com/NVlabs/DoRA/blob/main/commonsense_reasoning/peft/src/peft/tuners/dora.py) , it appears that the norms are actually computed over **output channels** (rows of the weight matrix). For example, in the `Linear` class's `train()` and `forward()` methods, the computation explicitly uses:

```python
norm_scale = ( self.weight_m_wdecomp.weight / (torch.linalg.norm(self.weight,dim=1)).unsqueeze(1) )
```

This directly computes the L2 norm over the rows of the weight matrix (`dim=1` for `[out_features, in_features]`), aligning with the methodology outlined in Section 3.1 of the DoRA paper. 

The author’s previous [comment](https://github.com/huggingface/peft/pull/1474#issuecomment-1964682771) highlights the importance of ensuring the magnitude vector only considers **post-dropout input**.

Given this, I wanted to double-check: Could the current implementation in `peft`—which computes norms over **input channels**—be diverging unintentionally from the original implementation and the paper?

Just want to understand if computing norms over input channels instead of output channels has no significant impact on results, such as training performance or stability.

Looking forward to your thoughts.

**Comment by arcteryox** [comment]
Thank you for taking the time to explain this in detail and for providing the code snippet to verify the behavior. You're absolutely right - now see where my misunderstanding arose.

My confusion stemmed from misinterpreting the conditional transpose as universally applied, rather than being specific to layers like Conv1D. Your clarification resolved this perfectly.

Sorry for my oversight in my initial report. The PEFT implementation is correct, and this bug report can be closed.

Thanks again.

**Comment by BenjaminBossan** [comment]
> Sorry for my oversight in my initial report. The PEFT implementation is correct, and this bug report can be closed.

No worries, we're always happy when others scrutinize the code base and report potential bugs.
